src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -0,0 +1,250 @@
+public:
+  spark_resource_adaptor(JNIEnv *env, rmm::mr::device_memory_resource *mr)
+      : resource{mr} {
+    if (env->GetJavaVM(&jvm) < 0) {
+      throw std::runtime_error("GetJavaVM failed");
+    }
+  }
+
+  rmm::mr::device_memory_resource *get_wrapped_resource() { return resource; }
+
+  bool supports_get_mem_info() const noexcept override { return resource->supports_get_mem_info(); }
+
+  bool supports_streams() const noexcept override { return resource->supports_streams(); }
+
+  void associate_thread_with_task(long thread_id, long task_id) {
+    std::scoped_lock lock(state_mutex);
+    if (threads.find(thread_id) != threads.end()) {
+      throw std::invalid_argument("a thread can only be added if it is in the unknown state");
+    }
+    threads.insert({thread_id, {false, false, task_id}});
+    auto recover = [this, thread_id]() {
+      threads.erase(thread_id);
+    };
+    {
+      rollback rb(recover);
+
+      auto tasks_it = task_to_threads.find(task_id);
+      if (tasks_it == task_to_threads.end()) {
+        task_to_threads.insert({task_id, {thread_id}});
+      } else {
+        tasks_it->second.insert(thread_id);
+      }
+    }
+  }
+
+  void remove_thread_association(long thread_id) {
+    std::scoped_lock lock(state_mutex);
+    auto threads_at = threads.find(thread_id);
+    if (threads_at != threads.end()) {
+      auto task_id = threads_at->second.task_id;
+      if (task_id > 0) {
+        auto task_at = task_to_threads.find(task_id);
+        if (task_at != task_to_threads.end()) {
+          task_at->second.erase(thread_id);
+        }
+      }
+      threads.erase(threads_at);
+    }
+  }
+
+  void force_retry_oom(long thread_id) {
+    std::scoped_lock lock(state_mutex);
+    auto threads_at = threads.find(thread_id);
+    if (threads_at != threads.end()) {
+      threads_at->second.retry_oom_injected = true;
+    } else {
+      throw std::invalid_argument("the thread is not associated with any task.");
+    }
+  }
+
+  void force_split_and_retry_oom(long thread_id) {
+    std::scoped_lock lock(state_mutex);
+    auto threads_at = threads.find(thread_id);
+    if (threads_at != threads.end()) {
+      threads_at->second.split_and_retry_oom_injected = true;
+    } else {
+      throw std::invalid_argument("the thread is not associated with any task.");
+    }
+  }
+
+
+private:
+  rmm::mr::device_memory_resource *const resource;
+  // The state mutex must be held when modifying the state of threads or tasks
+  // it must never be held when calling into the child resource or after returning
+  // from an operation.
+  std::mutex state_mutex;
+  std::map<long, thread_state> threads;
+  std::map<long, std::set<long>> task_to_threads;
+  JavaVM *jvm;
+
+  void *do_allocate(std::size_t num_bytes, rmm::cuda_stream_view stream) override {
+    auto tid = static_cast<long>(pthread_self());
+    {
+      // pre allocate checks
+      auto thread = threads.find(tid);
```

Comment:
this is calling into `threads` without the `state_mutex` held, I believe we should lock `state_mutex` here.